File: RepeatedDNASequences2.py
class Solution:
    def findRepeatedDnaSequences(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        hash_map = {}
        result = []
        for i in range(len(s)-9):
            v = 0
            for j in range(i, i+10):
                # (ord(s[j]) - ord('A') + 1)%5  A C G T 分别为1 3 2 0
                v  = v<<2 | (ord(s[j]) - ord('A') + 1)%5
            if v in hash_map:
                if hash_map[v] == 1:
                    result.append(s[i : i+10])
                hash_map[v] += 1
            else:
                hash_map[v] = 1
        return result
            
            
if __name__ == '__main__':
    s = raw_input()
    solution = Solution()
    print(solution.findRepeatedDnaSequences(s))

File: RepeatedDNASequences2.cpp
#include<iostream>
#include<string>
#include<vector>
#include<unordered_map>

using namespace std;

class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        
        vector<string> result;
        if(s.length()<10) return result;
        
        unordered_map<int, int> map;
        
        for(int i=0;i<s.length()-9;i++){
            int v = 0;
            for(int j = i;j<i+10;j++){
                v <<= 2;
                switch(s[j]){
                    case 'A':
                        break;
                    case 'C':
                        v |= 1;
                        break;
                    case 'G':
                        v |= 2;
                        break;
                    case 'T':
                        v |= 3;
                        break;
                }
            }
            if(map[v]++ == 1){
                result.push_back(s.substr(i, 10));
            }
            /*if(map.find(v) == map.end()){
                map[v] = 1;
            }else{
                if(map[v] == 1) result.push_back(s.substr(i, 10));
                map[v] ++;
            }*/
        }
        
        return result;
    }
};

int main(){
    string s;
    cin>>s;
    
    Solution *solution = new Solution();
    vector<string> result = solution->findRepeatedDnaSequences(s);
    for(int i=0;i<result.size();i++){
        cout<<result[i]<<endl;
    }
    
    return 0;
}
